fdb-extensions/src/main/java/com/apple/foundationdb/async/hnsw/StorageAdapter.java

```diff
@@ -0,0 +1,389 @@
+     * @param storageTransform an affine vector transformation operator that is used to transform the fetched vector
+     *        into the storage space that is currently being used
+     * @param layer the layer from which to fetch the node
+     * @param primaryKey the {@link Tuple} representing the primary key of the node to retrieve
+     * @return a non-null {@link CompletableFuture} which will complete with the fetched {@code Node<N>}.
+     */
+    @Nonnull
+    CompletableFuture<Node<N>> fetchNode(@Nonnull ReadTransaction readTransaction,
+                                         @Nonnull AffineOperator storageTransform,
+                                         int layer,
+                                         @Nonnull Tuple primaryKey);
+
+    /**
+     * Writes a node and its neighbor changes to the data store within a given transaction.
+     * <p>
+     * This method is responsible for persisting the state of a {@link Node} and applying any modifications to its
+     * neighboring nodes as defined in the {@code NeighborsChangeSet}. The entire operation is performed atomically as
+     * part of the provided {@link Transaction}.
+     * @param transaction the non-null transaction context for this write operation.
+     * @param quantizer the quantizer to use
+     * @param node the non-null node to be written to the data store.
+     * @param layer the layer index where the node resides.
+     * @param changeSet the non-null set of changes describing additions or removals of
+     *        neighbors for the given {@link Node}.
+     */
+    void writeNode(@Nonnull Transaction transaction, @Nonnull Quantizer quantizer, @Nonnull Node<N> node, int layer,
+                   @Nonnull NeighborsChangeSet<N> changeSet);
+
+    /**
+     * Scans a specified layer of the structure, returning an iterable sequence of nodes.
+     * <p>
+     * This method allows for paginated scanning of a layer. The scan can be started from the beginning of the layer by
+     * passing {@code null} for the {@code lastPrimaryKey}, or it can be resumed from a previous point by providing the
+     * key of the last item from the prior scan. The number of nodes returned is limited by {@code maxNumRead}.
+     *
+     * @param readTransaction the transaction to use for the read operation
+     * @param layer the index of the layer to scan
+     * @param lastPrimaryKey the primary key of the last node from a previous scan,
+     *        or {@code null} to start from the beginning of the layer
+     * @param maxNumRead the maximum number of nodes to return in this scan
+     * @return an {@link Iterable} that provides the nodes found in the specified layer range
+     */
+    Iterable<Node<N>> scanLayer(@Nonnull ReadTransaction readTransaction, int layer, @Nullable Tuple lastPrimaryKey,
+                                int maxNumRead);
+
+    /**
+     * Creates a {@code HalfRealVector} from a given {@code Tuple}.
```

Comment:
```suggestion
     * Creates a {@link RealVector} from a given {@code Tuple}.
```

I take it the `Half` bit is from an older version of this?

Reply:
yes -- fixed.